Add a test confirming that X509Certificate is trimmed

amcasey opened the pull request:
# Add a test confirming that X509Certificate is trimmed

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->



## Description

Add a test confirming that X509Certificate is trimmed when using `CreateSlimBuilder`.

amcasey committed:
Add a test confirming that X509Certificate is trimmed

...when using `CreateSlimBuilder`.

amcasey commented:
I hit a couple of surprises while authoring this:

1) I get a warning about JSON serialization.  I think this indicates that the feature switch for using reflection by default is enabled, which is not the case in "real" native AOT scenarios.
```
src\Http\Http.Extensions\src\JsonOptions.cs(39): Trim analysis warning IL2026: Microsoft.AspNetCore.Http.Json.JsonOptions.CreateDefaultTypeResolver(): Using member 'System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.DefaultJsonTypeInfoResolver()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. JSON serialization and deserialization might require types that cannot be statically a nalyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext, or make sure all of the required types are preserved.
src\Http\Http.Extensions\src\JsonOptions.cs(39): AOT analysis warning IL3050: Microsoft.AspNetCore.Http.Json.JsonOptions.CreateDefaultTypeResolver(): Using member 'System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.DefaultJsonTypeInfoResolver()' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. JSON serialization and deserialization might require types that cannot be statically analyzed and might need runtime code generation. Use System.Text.Json source generation for native AOT applications.
```

2) I was originally planning to sketch out a more complete app, but I got errors when I tried to use `MapGet`
```
artifacts\bin\trimmingTests\projects\Microsoft.AspNetCore.NativeAotTests\SlimBuilderDoesNotDependOnX509Test\SlimBuilderDoesNotDependOnX509Test.cs(12,1): error IL3050: Using member 'Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapGet(IEndpointRouteBuilder, String, Delegate)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. This API may perform reflection on the supplied delegate and its parameters. These types may require generated code and aren't compatible with native AOT applications.
artifacts\bin\trimmingTests\projects\Microsoft.AspNetCore.NativeAotTests\SlimBuilderDoesNotDependOnX509Test\SlimBuilderDoesNotDependOnX509Test.cs(12,1): error IL2026: Using member 'Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapGet(IEndpointRouteBuilder, String, Delegate)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. This API may perform reflection on the supplied delegate and its parameters. These types may be trimmed if not directly referenced.
artifacts\bin\trimmingTests\projects\Microsoft.AspNetCore.NativeAotTests\SlimBuilderDoesNotDependOnX509Test\SlimBuilderDoesNotDependOnX509Test.cs(12,1): error IL3050: Using member 'Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapGet(IEndpointRouteBuilder, String, Delegate)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. This API may perform reflection on the supplied delegate and its parameters. These types may require generated code and aren't compatible with native AOT applications.
artifacts\bin\trimmingTests\projects\Microsoft.AspNetCore.NativeAotTests\SlimBuilderDoesNotDependOnX509Test\SlimBuilderDoesNotDependOnX509Test.cs(12,1): error IL2026: Using member 'Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions.MapGet(IEndpointRouteBuilder, String, Delegate)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. This API may perform reflection on the supplied delegate and its parameters. These types may be trimmed if not directly referenced.
```

@eerhardt or @captainsafia might have thoughts on how the test environment differs from publishing the AOT template.

JamesNK commented on src/DefaultBuilder/test/Microsoft.AspNetCore.NativeAotTests/SlimBuilderDoesNotDependOnX509Test.cs:
```diff
@@ -0,0 +1,23 @@
+var builder = WebApplication.CreateSlimBuilder(args);
+_ = builder.Build();
+
+if (GetType("System.Security.Cryptography", "System.Security.Cryptography.X509Certificates.X509Certificate") is not null) {
```
Here's an idea, and you can feel free to tell me to go away:

A test like this could pass because `X509Certificate` has been trimmed away, but it could also pass because of a typo in the type name.

Can we verify that `X509Certificate` is present when the regular `CreateBuilder` is used? Either:
* Have a separate console app that uses `CreateBuilder` and share this assert between them
* Or have a compiler constant + `#ifdef` and run the file twice with different compiler constant values (this would probably require complex build changes).

amcasey replied:
I used `CreateBuilder` as part of my local validation that the test would catch regressions, but it actually fails to build because others parts of `CreateBuilder` aren't AOT compatible. :cry:

amcasey replied:
I wonder if I could write that as a trimming test though...

JamesNK replied:
`CreateBuilder `works with AOT. The gRPC template uses it and it builds with AOT without warnings: https://github.com/dotnet/aspnetcore/blob/c5a9c5973bf735f1bb9d611cd50c796f6c32557e/src/ProjectTemplates/Web.ProjectTemplates/content/GrpcService-CSharp/Program.cs#L3

amcasey replied:
Alright, then I just need to figure out the combination of feature switches that will make that true in the test.

amcasey replied:
It's still getting trimmed with `CreateBuilder`.  Rather than investigate deeply, I added a test confirming it's preserved when `UseHttps` is invoked.  Now let's confirm I didn't make it depend on having a dev cert...

amcasey committed:
Disable IsReflectionEnabledByDefault to eliminate warning

amcasey committed:
Switch to trimming tests

...since we're testing trimming behavior and add a test in the other direction (i.e. `UseHttps` does cause `X509Certificate` to be preserved).

amcasey committed:
Hack around the need for a certificate

JamesNK commented on src/DefaultBuilder/test/Microsoft.AspNetCore.TrimmingTests/UseHttpsDoesDependOnX509Test.cs:
```diff
@@ -0,0 +1,23 @@
+    serverOptions.ListenLocalhost(5000, listenOptions =>
+    {
+#pragma warning disable SYSLIB0026 // The constructor obsolete but we're not actually going to use the cert
+        listenOptions.UseHttps(new X509Certificate2());
```
You're directly referencing the cert. The assert that the cert type is loaded doesn't make sense anymore.

Change the `GetType` to be something internal to Kestrel? Or wrap `builder.Build()` in a try-catch and comment that the error is expected because a dev cert isn't present.

amcasey replied:
:facepalm:

In my defence, the only purpose of the test is to confirm the helper can detect a particular preserved type, which it still does.  But I like the try-catch idea.

amcasey committed:
Mitigate missing dev cert in a way that doesn't invalidate the test

JamesNK reviewed (approved):
LGTM. I have yet to write one of these, so I don't know all the ins and outs. Wait for @eerhardt to double-check its correctness.

amcasey commented:
The test failure is clearly me, but it's not obvious to me what to do about it.  I'm guessing it has something to do with the way we're depending on aspnetcore via source, rather than binaries.
```
CSC : error CS0006: Metadata file '/mnt/vss/_work/1/s/artifacts/obj/Microsoft.AspNetCore.DataProtection.Extensions/Release/net8.0/ref/Microsoft.AspNetCore.DataProtection.Extensions.dll' could not be found [/mnt/vss/_work/1/s/artifacts/bin/trimmingTests/projects/Microsoft.AspNetCore.TrimmingTests/SlimBuilderDoesNotDependOnX509Test/project.csproj]
```

eerhardt commented on src/DefaultBuilder/test/Microsoft.AspNetCore.TrimmingTests/Microsoft.AspNetCore.TrimmingTests.proj:
```diff
@@ -0,0 +1,14 @@
+
+  <ItemGroup>
+    <TestConsoleAppSourceFiles Include="SlimBuilderDoesNotDependOnX509Test.cs">
+      <DisabledFeatureSwitches>System.Text.Json.JsonSerializer.IsReflectionEnabledByDefault</DisabledFeatureSwitches>
```
What was happening when you didn't set this? You were getting trim warnings during publish and that broke your test?

eerhardt replied:
Note that the `NativeAotTests.proj` doesn't need to do this because `IsReflectionEnabledByDefault` is defaulted to `false` when `PublishAot=true`. https://github.com/dotnet/runtime/issues/84378 is tracking doing it for `PublishTrimmed` as well.

amcasey replied:
When it lived in NativeAotTests, I was getting a warning about it.  In TrimmingTests, it was an error, which broke build.

amcasey replied:
For posterity, the error was
```
src\Http\Http.Extensions\src\JsonOptions.cs(39,12): Trim analysis error IL2026: Microsoft.AspNetCore.Http.Json.JsonOptions.CreateDefaultTypeResolver(): Using member 'System.Text.Json.Serialization.Metadata.DefaultJsonTypeInfoResolver.DefaultJsonTypeInfoResolver()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo or JsonSerializerContext, or make sure all of the required types are preserved.
```

eerhardt commented on src/DefaultBuilder/test/Microsoft.AspNetCore.TrimmingTests/SlimBuilderDoesNotDependOnX509Test.cs:
```diff
@@ -0,0 +1,10 @@
+var app = builder.Build();
+
+if (X509Utilities.HasCertificateType) {
+    return 1;
```
(nit) The convention is to use negative numbers. That way if you ever get more than 99 checks, you don't have to skip over 100. And it is easier to intuit that a negative exit code means something went wrong.

```suggestion
    return -1;
```

amcasey replied:
Ah, the classic hundred-failure-modes unit test...

eerhardt commented on src/DefaultBuilder/test/Microsoft.AspNetCore.TrimmingTests/SlimBuilderDoesNotDependOnX509Test.cs:
```diff
@@ -0,0 +1,10 @@
+using Microsoft.AspNetCore.Builder;
```
Should add copyright to all code files.

amcasey replied:
Oops, lost that when I started building in a scratch project on the side.  I got lint warnings about not using ConfigureAwait, but not about the missing header?

eerhardt replied:
> lint warnings

lol

eerhardt commented on src/DefaultBuilder/test/Microsoft.AspNetCore.TrimmingTests/X509Utilities.cs:
```diff
@@ -0,0 +1,20 @@
+    public static bool HasCertificateType =>
+        GetType("System.Security.Cryptography", "System.Security.Cryptography.X509Certificates.X509Certificate") is not null;
+
+    [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+        Justification = "Returning null when the type is unreferenced is desirable")]
+    private static Type? GetType(string assemblyName, string typeName) {
+        var assembly = AppDomain.CurrentDomain.GetAssemblies().SingleOrDefault(a => a.GetName().Name == assemblyName);
+        return assembly?.GetType(typeName);
```
Why not use `Type.GetType` here?

amcasey replied:
It didn't work.  Happy to learn why and use a more appropriate API.

eerhardt replied:
What didn't work?

amcasey replied:
AFAICT, it returned null for all `Microsoft.AspNetCore` types, regardless of whether or not they were preserved.  (It did consistently find `System.Threading.Tasks.Task` though.)

amcasey replied:
The [runtime tests](https://github.com/dotnet/runtime/blob/e4743ec9474728f2b4e120fd70a3f729e9204c32/src/libraries/System.Text.Json/tests/System.Text.Json.Tests/TrimmingTests/IsReflectionEnabledByDefaultFalse.cs#L65) also appear to avoid `Type.GetType`, though that may be for some other reason.

amcasey replied:
@eerhardt pointed out offline that it wasn't working because I wasn't passing the assembly name.  Fixed.

amcasey committed:
Delay-build trimming test projects

amcasey committed:
Add copyright headers

amcasey commented:
> So to fix this for just your PR, you can add a <RequiresDelayedBuild Include=" line to that file for your new test project.

Ah, just like the line for the NativeAotTests project I copied. 😆 

Fixed it on my box, hopefully in CI too.

amcasey committed:
Use negative exit codes in error scenarios

eerhardt reviewed (approved):
Looks great! Thanks for the good work.

amcasey committed:
Simplify type lookup

mitchdenny commented:
@eerhardt I notice this failed on macOS. Is there some way to get more detailed output in these cases?

eerhardt commented:
Re-running to see if it passes on retry.  (disabled auto-merge in case it does pass)

eerhardt commented:
> [Log](https://dev.azure.com/dnceng-public/public/_build/results?buildId=279229&view=logs&j=f43b187f-6c15-531c-a59b-e4fba4e9ed20&t=98a09275-a4dd-593e-c21a-1f8bea3be5b1&l=3137) says:
> 
> ```
> The Command /Users/runner/work/1/s/artifacts/bin/trimmingTests/projects/Microsoft.AspNetCore.TrimmingTests/SlimBuilderDoesNotDependOnX509Test/bin/Release/net8.0/osx-x64/publish/project returned a non-success exit code: 255.
> ```
> 
> Why that would happen... I'm not sure.

According to https://www.redhat.com/sysadmin/exit-codes-demystified, returning an exit code greater than 256 will subtract 256. So exit code 255 means the app returned `-1`. There must be something on macOS that is still preserving `X509Certificate`.

amcasey commented:
That's odd.  I think the mac leg was passing (somehow) before I added the delay-build entry.  I just assumed these tests weren't running on mac.

amcasey commented:
It passed on re-run, so I don't think it can have been about preserving the dependency?

eerhardt commented:
Maybe let's run it again to see if it passes. If it does, we can merge it and monitor if it fails in CI again. We can always disable it on macOS pretty quickly if it is flakey.

amcasey commented:
Oh, I didn't want to create noise, but I have a draft PR [here](https://github.com/dotnet/aspnetcore/pull/48332) running it repeatedly.  I believe it's failing. 🙃

amcasey commented:
The tests pass when I revert back to my wacky way of looking up the types (i.e. not using `Type.GetType`).

amcasey commented:
Force push is a trivial rebase on latest `main`.  The test now fails locally as well (Eric found https://github.com/dotnet/aspnetcore/pull/48293/files#diff-946dd406190b49f4dd925f5a414d3f689d966b4c8044344ebb76f89952be5f9bR79).  Fix pending.

amcasey committed:
Allow X509Certificate to be preserved if it has no members

amcasey commented:
This CI pass is legit (i.e. zero errors). :smile:

amcasey commented:
I'm going to call this good and look into the build status issue separately.

eerhardt commented on src/DefaultBuilder/test/Microsoft.AspNetCore.TrimmingTests/X509Utilities.cs:
```diff
@@ -0,0 +1,40 @@
+        {
+            var certificateType = GetType("System.Security.Cryptography", "System.Security.Cryptography.X509Certificates.X509Certificate");
+
+            // We're checking for members, rather than just the presence of the type,
+            // because Debugger Display types may reference it without actually
+            // causing a meaningful binary size increase.
+            return certificateType is not null && GetMembers(certificateType).Any();
```
Another approach we could take here is to disable the feature switch `System.Diagnostics.Debugger.IsSupported` (`DisabledFeatureSwitches` in the .proj file) and ensure this type is fully trimmed. This will get us closer to what AOT does, because that feature switch is disabled by default for PublishAot.

https://github.com/dotnet/runtime/blob/ce689d9189d549e249f4bb5e3d9aeef322464bfb/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L256-L257

amcasey replied:
I thought about that but worried it would hide problems.  I could add a separate test for that flag, if you think it's worthwhile.

eerhardt replied:
> but worried it would hide problems

What kind of problems?

This mode would more closely align with the Native AOT behavior. So it seems the more preferrable to me. The fact that we are rooting types for Debugging published apps is less desirable to verify (IMO).
